### DISCUSSION TITLE
Angular support/example

### DISCUSSION THREAD

**Discussion by johanndev** [facebook/stylex#97] [General]
**Is your feature request related to a problem? Please describe.**

I want to try stylex in an angular project. As angular is explicitly [mentioned](https://stylexjs.com/docs/learn/#authoring-ui-in-javascript) in the introduction section of the docs, I want to give it a go.

**Describe a solution you'd like**

Provide instructions on how to integrate stylex with angular. The [instructions for integrating with webpack](https://stylexjs.com/docs/learn/installation/#production) already provide the base for the integration, but comprehensive documentation on how to apply the stylex API is missing.

**Describe alternatives you've considered**

**Additional context**

**Comment by kevintyj** [comment]
The JS/API usage docs are located here: https://stylexjs.com/docs/api/javascript/props/

**Comment by johanndev** [comment]
Thanks @kevintyj!

After a little bit of tinkering, I arrived at this first working example:
```ts
import { bootstrapApplication } from '@angular/platform-browser';
import { Component, isDevMode } from '@angular/core';
import inject from '@stylexjs/dev-runtime';
import { CommonModule } from '@angular/common';
import stylex, { props } from '@stylexjs/stylex';

if (isDevMode()) {
  inject({
    classNamePrefix: 'x',
    dev: true,
    test: false,
    useRemForFontSize: true,
    runtimeInjection: true,
    styleResolution: 'application-order',
  });
}

@Component({
  selector: 'app-root',
  standalone: true,
  imports: [CommonModule],
  template: `
    <p [ngClass]="classBinding">Test</p>
  `,
})
class AppComponent {
  classBinding: string | undefined;

  constructor() {
    const styles = stylex.create({
      container: {
        border: '1px dashed black;'
      },
      text: {
        color: 'red',
      },
    });
    const propsRes = props(styles.container, styles.text);
    this.classBinding = propsRes.className;
  }
}

bootstrapApplication(AppComponent).catch((err) => console.error(err));

```

I'm pretty sure this isn't the 'ideal' way of using stylex in angular, but it's a start :)

**Comment by nmn** [comment]
@johanndev This is close, but it should be more like this:

```ts
import { bootstrapApplication } from '@angular/platform-browser';
import { Component, isDevMode } from '@angular/core';
import inject from '@stylexjs/dev-runtime';
import { CommonModule } from '@angular/common';
import * as stylex from '@stylexjs/stylex';

if (isDevMode()) {
  inject({ dev: true, useRemForFontSize: true });
}

const styles = stylex.create({
  container: {
    borderWidth: 1,
    borderStyle: 'dashed',
    borderColor: 'black',
  },
  text: {
    color: 'red',
  },
});

@Component({
  selector: 'app-root',
  standalone: true,
  imports: [CommonModule],
  template: `
    <p [ngClass]="classBinding">Test</p>
  `,
})
class AppComponent {
  classBinding: string | undefined;
  styleBinding: string | undefined;

  constructor() {
    const {className, style} = stylex.props(styles.container, styles.text);
    this.classBinding = propsRes.className;
    // Does Angular support objects for inline styles?
    this.styleBinding = style && Object.entries(style)
      .map((pair) => pair.join(':'))
      .join(';');
  }
}

bootstrapApplication(AppComponent).catch((err) => console.error(err));
```

**Notes:**

1. It's important to key the `stylex.create` call at the top level of the module 
so that the generated objects of classNames have stable references for memoization.

2. We don't recommend using shorthand properties with multiple space-separated values.
   (e.g. I split up the border shorthand from your example)


**Questions:**

1. Is there a way in Angular (perhaps in the newest release?) to bind properties directly to the template without first assigning a class property?

2. The Babel plugin should *just* work with this. Perhaps someone with knowledge of    Angular's type bundler toolchain can help?

**Comment by kevintyj** [comment]
Haven't tried StyleX with Angular, but as of directly binding to the template, there is no way to achieve that as far as I am aware but we could use a directive to abstract this out like so:

```ts
@Directive({
	selector: '[sx]',
	standalone: true, // If you are using standalone
	host: {
		'[class]': '_generateClass',
		'[style]': '_generatedStyle',
	},
})
export class StyleXDirective {
	private readonly _sxStyles: StyleXStyles;
	@Input('sx')
	set class(sxObj: StyleXStyles) {
		this._sxStyles = sxObj;
	}

	protected _generateClass: string | undefined;
	protected _generatedStyle: string | undefined;

	constructor() {
		const {className, style} = stylex.props(styles.container, styles.text);
		this._generateClass = propsRes.className;
		// Does Angular support objects for inline styles?
		this._generatedStyle = style && Object.entries(style)
			.map((pair) => pair.join(':'))
			.join(';');
	}
}
```

I have no clue if this works though, I also directly took the code form @nmn for the part that actually takes the styles props.

**Comment by johanndev** [comment]
I was just going to post a similar approach via directive:
```ts
import { bootstrapApplication } from '@angular/platform-browser';
import {
  Component,
  Directive,
  HostBinding,
  Input,
  isDevMode,
} from '@angular/core';
import { CommonModule } from '@angular/common';
import stylex from '@stylexjs/stylex';
import injestylexInject from '@stylexjs/dev-runtime';

if (isDevMode()) {
  injestylexInject({
    classNamePrefix: 'x',
    dev: true,
    test: false,
    useRemForFontSize: true,
    runtimeInjection: true,
    styleResolution: 'application-order',
  });
}

const styles = stylex.create({
  container: {
    borderWidth: 1,
    borderStyle: 'dashed',
    borderColor: 'black',
  },
  text: {
    color: 'blue',
  },
});

@Directive({
  selector: '[stylexProps]',
  standalone: true,
})
class StylexClassDirective {
  @Input({ required: true }) set stylexProps(
    stylexProps: ReturnType<typeof stylex.create>
  ) {
    this.props = stylex.props(stylexProps['container'], stylexProps['text']);
  }
  private props: ReturnType<typeof stylex.props> | undefined;
  @HostBinding('class')
  get className() {
    return this.props?.className;
  }
  get style() {
    return this.props?.style;
  }
}

@Component({
  selector: 'app-root',
  standalone: true,
  imports: [CommonModule, StylexClassDirective],
  template: ` <p [stylexProps]="localStyles">Styles bound via Class</p> `,
})
class AppComponent {
  protected localStyles = styles;
}

bootstrapApplication(AppComponent).catch((err) => console.error(err));

```

**Comment by johanndev** [comment]
Regarding the comment:
> // Does Angular support objects for inline styles?

Angular provides the [`ngStyle` directive](https://angular.io/api/common/NgStyle) which can take an object containing key-value pairs. In the example above, `style` returned from `style.props(...)` always remains `undefined` for me, so I wasn't able to test it.

**Comment by necolas** [comment]
We don't use Angular at Meta and are unlikely to maintain an example for it, especially if it spawns requests for every other framework to get an example.

StyleX has just been released, so I suspect that in the near future members of communities like Angular will publish posts or examples of how best to integrate StyleX, and that is probably the best place for it.

**Comment by necolas** [comment]
Going to move this over to "discussions"

**Comment by nmn** [comment]
Seeing the examples so far, unless a custom directive can be added or something, I don't see a nice way of using StyleX with Angular. 

- Defining styles with names
- Assinging them to the class as properties
- reference those properties by name in the template *TWICE*

Would Angular developers find this acceptable? This feels like an unacceptable level of indirection to me. 

Unless the angular developers here feel otherwise, I think we should remove the `Angular` mention from our docs and close this discussion.

**Reply by kevintyj** [reply]
If you use a custom stylex directive / pipes, I dont think it would be too different from the implementation with React (not referencing class as props in the template twice). I remember using stitches with Angular and it had a very similar approach of just appending classes to the `[ngClass]` selector. 

```html
// sx or stylex directive handles assigning the class names internally without having to do that with every component. 
<div sx={...props.something}>Some Text</div>
```

I'm not quite understanding what you mean by assigning them to the class as props then referencing the props in the template twice, you could just define the styles within the component/module then just directly reference the styles (in the template) using the custom directive. But also I haven't tested this approach yet!

**Reply by nmn** [reply]
Can you show a more complete example? My understanding is that the template string in template has to be static and any dynamic values need to be put on the class and read from there.

**Reply by kevintyj** [reply]
I have not tested the code below but using a pattern like this should work:
```ts
// Custom Angular Directive from the answer above
@Directive({
	selector: '[sx]',
	standalone: true,
	host: {
		'[class]': '_generateClass',
		'[style]': '_generatedStyle',
	},
})
export class StyleXDirective {
	private readonly _sxStyles: StyleXStyles;
	@Input('sx')
	set class(sxObj) {
		this._sxStyles = sxObj;
	}

	protected _generateClass: string | undefined;
	protected _generatedStyle: string | undefined;

	constructor() {
		const {className, style} = stylex.props(sxObj);
		this._generateClass = propsRes.className;
	}
}
```
This is the custom directive part to abstract the generation of styles and binding to the correct props of the object, the constructor definitely has to be re-written and is just taken from answers above.
Then inside a another component:
```ts
@Component({
	standalone: true,
	imports: [CommonModule],
	templateUrl: `<div [sx]={styleXObj.coolComponent}></div>`,
})
export class SomeSampleComponent extends StyleXDirective {
	protected styleXObj = stylex.create({
		coolComponent: {
			color: 'red'
		}
	})
}
```
Something like this is what I am thinking... Might be completely off the mark, but would have to carefully think about how types are handled in these cases

**Reply by nmn** [reply]
Even if this works, you'd need to at least do this:

```ts
const styles = stylex.create({
  coolComponent: {
    color: 'red'
  }
});

@Component({
	standalone: true,
	imports: [CommonModule],
	templateUrl: `<div [sx]={styleXObj}></div>`,
})
export class SomeSampleComponent extends StyleXDirective {
	protected styleXObj = styles.coolComponent;
}
```

And do something like this to accept style props:

```ts
@Component({
  standalone: true,
  imports: [CommonModule],
  templateUrl: `<div [sx]={styleXObj}></div>`,
})
export class SomeSampleComponent extends StyleXDirective {
  // How do you accept `style` as a "prop"?
  protected styleXObj = [styles.coolComponent, style];
}
```

Is this a good experience? Personally, I'd want something like this to work:

```ts
@Component({
  // ...
  templateUrl: `<div [sx]=${[styles.coolComponent, style]}></div>`,
})
export class SomeSampleComponent extends StyleXDirective {}
```

**Reply by SpeedoPasanen** [reply]
Your templates are totally wrong so I'm not sure I get what you mean, but yes you can pass an array as input to a directive or component. You don't use template string interpolation in Angular.

Say `<div [sx]="[item1, item2]">` to pass an array.

**Reply by SpeedoPasanen** [reply]
Oh sorry, you mean you don't want to set those properties to the component instance but pass variables them from the outside of component code.... Yeah that's not possible because the component and it's template is compiled AOT.

**Reply by nmn** [reply]
Yeah, I'm not familiar with Angular and it just feels a bit off. As someone who's familiar with Angular does any of the *possibly* examples feel like something you'd be interested in?
